src/lib.rs

```diff
@@ -24,10 +24,30 @@ pub use types::{Alignment, FmtError, Result, Sign};
 fmtint!(u8 i8 u16 i16 u32 i32 u64 i64 usize isize);
 fmtfloat!(f32 f64);
 
+pub trait TypeFormatting{
+    fn do_format(&self,f:&mut Formatter) -> Result<()>;
+}
+
+impl TypeFormatting for &str {
+    fn do_format(&self, f: &mut Formatter) -> Result<()> {
+        f.str(self)
+    }
+}
+
+impl TypeFormatting for String {
+    fn do_format(&self, f: &mut Formatter) -> Result<()> {
+        f.str(self)
+    }
+}
+
+fmttype!(u8 i8 u16 i16 u32 i32 u64 i64 usize isize);
+fmttype!(f32 f64);
+
 /// Rust-style format a string given a `HashMap` of the variables.
-pub fn strfmt<K, T: fmt::Display>(fmtstr: &str, vars: &HashMap<K, T>) -> Result<String>
```

Comment:
So this is changing the type API of the library, right?

With these changes, the library will no longer allow just any type which implements `Display`, folks will have to specifically implement `TypeFormatting` for their custom types, correct?

Reply:
Correct, but only if you would like to pass types to the Formatter where we don't know which Formatter should resolve this. I see what you want here and I got an idea of how to fix this, ill try to update as soon as I get more time.

Currently I am unavailable until 1 Aug.

Reply:
I'm wondering if there is a way to auto-implement `TypeFormatting` for any type which implements `Display`?

https://stackoverflow.com/a/29256897/1036670

```
impl<D> TypeFormatting for D
where
    D: Display,
{
    fn do_format(&self, f:&mut Formatter) -> Result<()> {
        f.str(format!(self))
    }
}
```

Something like that... my rust is a bit... rusty.

Reply:
If we did the above, I _believe_ we could get rid of `strfmt_ext` but also not break backwards compatibility? If you think I'm right I'd appreciate some tests to prove it!

And thanks for doing this work, I really appreciate it. I just want to make sure the API is as good as it can be.

Reply:
That's the problem, this generic trait implementation would also implement the behavior for other types, like the already implemented stuff for f64,f32,u8,i8.... I don't know any trait that allows us to disguise between primitive and non-primitive, as well as the question mentioned in your StackOverflow Discussion is "why should Rust?".

And to say this clearly, I appreciate this review. I'm total with you, the API should be as good as it can be.